Extend the custom Jump List API provided by Electron

enlight opened the pull request:
Add `app.setJumpList(callback)` which makes it possible to fully customize the Jump List of an Electron app, specifically it allows to:
- Add tasks to the standard `Tasks` category.
- Add separators to the standard `Tasks` category.
- Add custom categories containing tasks and file links.
- Add system managed Recent/Frequent categories.
- Remove the custom Jump List.

`app.setUserTasks()` should probably be deprecated at some point.

zcbenz commented:
Current API looks inconsistent with other APIs, usually `callback` is used for event handlers or request callbacks, it is weird to use `callback` for setting parameter synchronously.

I under understand it is for providing the `minItems, removedItems`, but since very few developers would really use them, maybe we should make the `setJumpList` API accept one object parameters, and add another `getJumpListSettings` API to return `minItems, removedItems`?

enlight commented:
Yeah, I agree the API is inconsistent, I'll make the suggested change.

enlight commented:
Updated the PR to add `getJumpListSettings()`.

zcbenz commented on atom/browser/browser.h:
```diff
@@ -29,6 +29,22 @@ namespace gfx {
 
+#if defined(OS_WIN)
+namespace atom {
+enum class JumpListResult : int;
```
Since the `SetJumpList` is a public API of `Browser`, the `JumpListResult` should also be public too. So the complete definition of `JumpListResult` should be placed in `browser.h`.

zcbenz commented on atom/browser/browser.h:
```diff
@@ -29,6 +29,22 @@ namespace gfx {
+namespace mate {
+
+template<>
+struct Converter<atom::JumpListResult> {
```
The converters for a type should be placed under the `_api.cc` files that use them, for this one, it should be put in `atom_api_app.cc`.

zcbenz commented on atom/browser/browser.h:
```diff
@@ -158,6 +174,12 @@ class Browser : public WindowListObserver {
   bool SetUserTasks(const std::vector<UserTask>& tasks);
 
+  // Get the current Jump List settings.
+  v8::Local<v8::Value> GetJumpListSettings(mate::Arguments* args);
```
The APIs under `atom::Browser` are also meant to be used by other parts of Electron code apart from V8 bindings, so methods under this class should not have dependency of `V8`. If an API is only exposed in JavaScript, it should be put in the `api::` classes directly. This is for the good of having a clean architecture.

So we should either make `Browser::GetJumpListSettings` and `Browser::SetJumpList` independent from V8, or just put them under `api::App`. And the `browser.cc` should not include any V8 code.

A few places are violating this rule and escaped review, and we will fix them.

enlight replied:
Ah OK, I wasn't clear on the distinction between Browser and App, thanks for explaining it.

enlight commented:
Rebased on master and moved `setJumpList` and `getJumpListSettings` to `atom_app_api.h/cc` since they're mostly just glue code. I've left `setUserTasks` in `Browser` for now, if you want it moved to `App` I can submit a separate PR.

enlight commented:
Hmm, no idea why those builds on Travis failed.

zcbenz commented on atom/browser/api/atom_api_app.cc:
```diff
@@ -523,6 +736,63 @@ void App::OnCertificateManagerModelCreated(
 #endif
 
+#if defined(OS_WIN)
+v8::Local<v8::Value> App::GetJumpListSettings(mate::Arguments* args) {
```
Since the code is not using members of `mate::Arguments`, this parameter can be omitted and the isolate can be got from `isolate()` member.

The method can also return `mate::Dictionary` directly.

enlight replied:
Returning `mate::Dictionary` directly would be inconsistent with the rest of Electron's bindings code.

zcbenz commented on atom/browser/api/atom_api_app.h:
```diff
@@ -24,9 +24,21 @@ namespace base {
 }
 
+namespace atom {
+enum class JumpListResult : int;
```
This can be moved to inside `namespace atom` below.

zcbenz commented on atom/browser/api/atom_api_app.h:
```diff
@@ -24,9 +24,21 @@ namespace base {
+
+#if defined(OS_WIN)
+template<>
+struct Converter<atom::JumpListResult> {
```
This declaration seems redundant.

zcbenz commented on atom/browser/ui/win/jump_list.cc:
```diff
@@ -0,0 +1,331 @@
+  DCHECK(shell_item);
+  DCHECK(file_name);
+
+  wchar_t* file_name_buffer = nullptr;
```
There is `base::win::ScopedCoMem` to avoid manual memory freeing.

zcbenz commented on atom/browser/ui/win/jump_list.cc:
```diff
@@ -0,0 +1,331 @@
+
+  item->type = JumpListItem::Type::TASK;
+  wchar_t path[MAX_PATH];
+  constexpr size_t max_path = std::extent<decltype(path)>::value;
```
While this is nice, in this simple case using the `arraysize` macro would make the code more easy to read, `arraysize` is used very widely in Chromium so everyone knows what `arraysize` means, and it is shorter to use.

zcbenz commented on atom/browser/ui/win/jump_list.cc:
```diff
@@ -0,0 +1,331 @@
+  base::win::ScopedPropVariant prop;
+  if (SUCCEEDED(property_store->GetValue(PKEY_Link_Arguments, prop.Receive()))
+      && (prop.get().vt == VT_LPWSTR)) {
+    item->arguments = base::string16(prop.get().pwszVal);
```
I didn't test but the `base::string16` wrapper seems unnecessary.

zcbenz commented on atom/browser/ui/win/jump_list.cc:
```diff
@@ -0,0 +1,331 @@
+      && (prop.get().vt == VT_LPWSTR)) {
+    item->arguments = base::string16(prop.get().pwszVal);
+  } else {
+    item->arguments.clear();
```
The `clear` calls in this function are redundant, since the members of `item` are empty or assigned to 0 by default.

zcbenz commented on atom/browser/ui/win/jump_list.cc:
```diff
@@ -0,0 +1,331 @@
+    IShellItem* shell_item;
+    IShellLink* shell_link;
+    for (UINT i = 0; i < removed_count; ++i) {
+      if (SUCCEEDED(in->GetAt(i, IID_PPV_ARGS(&shell_item)))) {
```
The standard way for this in Electron is to protect the item with `base::win::ScopedComPtr`, so there is no manual release needed.

```c++
base::win::ScopedComPtr<IShellItem> shell_item;
in->GetAt(i, IID_PPV_ARGS(shell_item.Receive()));
```

enlight replied:
The same could be done with `CComPtr` so I don't see the need to use `base::win::ScopedComPtr` here unless you want me to replace all the other `CComPtr`s in this file. I didn't use `CComPtr` in this case because the loop is very simple and constructing two `CComPtr`s every iteration didn't seem like it was worth the effort.

zcbenz commented on atom/browser/ui/win/jump_list.cc:
```diff
@@ -0,0 +1,331 @@
+namespace atom {
+
+bool JumpList::CreateDestinationList() {
+  return SUCCEEDED(destinations_.CoCreateInstance(CLSID_DestinationList));
```
Putting this call in constructor can make code simpler.

zcbenz commented on docs/api/app.md:
```diff
@@ -527,6 +527,151 @@ Adds `tasks` to the [Tasks][tasks] category of the JumpList on Windows.
 Returns `true` when the call succeeded, otherwise returns `false`.
 
+**Note:** If you'd like to customize the Jump List even more use
+`app.setJumpList(callback)` instead.
```
This should be updated.

zcbenz commented on docs/api/app.md:
```diff
@@ -527,6 +527,151 @@ Adds `tasks` to the [Tasks][tasks] category of the JumpList on Windows.
+* `name` String - Must be set if `type` is `"custom"`, otherwise it should be
+  omitted.
+* `items` Array - Array of `JumpListItem` objects if `type` is `"tasks"` or
+          `"custom"`, otherwise it should be omitted.
```
This line is not aligned.

zcbenz commented on docs/api/app.md:
```diff
@@ -527,6 +527,151 @@ Adds `tasks` to the [Tasks][tasks] category of the JumpList on Windows.
+* `program` String - Path of the program to execute, usually you should
+  specify `process.execPath` which opens the current program. Should only be
+  set if `type` is `"task"`.
+* `arguments` String - The command line arguments when `program` is
```
Can you rename this to `args`? It was a wrong decision to use `arguments` since it is a keyword of JavaScript and all other places are using `args`, it is good a chance to fix this in the new API.

enlight commented:
OK, I've made all the changes you've requested, except where you left a :+1: on my comments (which I assume means you agree with me on those points). Here's the list of changes since the previous version:

- Remove redundant forward declaration of `Converter<atom::JumpListResult>`.
- Remove `mate::Arguments` parameter from `App::GetJumpListSettings()`.
- Use `base::win::ScopedCoMem` in `GetShellItemFileName()`.
- Replace `std::extent<decltype(path)>::value` with the more readable `arraysize(path)`.
- Remove redundant calls to `base::string16` in `ConvertShellLinkToJumpListItem()`.
- Remove redundant calls to `clear()` in the JumpListItem `mate` converter.
- Merge `JumpList::CreateDestinationList()` into the `JumpList` constructor.
- Rename the `arguments` property of the Jump List item object to `args` in the public API since the former is a JS keyword.
- Clean up the docs.

P.S. Unfortunately GitHub doesn't send me notifications when you leave a :+1:, so next time, make some noise when you review :)

zeke commented on docs/api/app.md:
```diff
@@ -527,6 +527,151 @@ Adds `tasks` to the [Tasks][tasks] category of the JumpList on Windows.
+  Jump List (for a more detailed description of this value see the
+  [MSDN docs][JumpListBeginListMSDN]).
+* `removedItems` Array - Array of `JumpListItem` objects that correspond to
+  items that the user has explicitely removed from custom categories in the
```
explicitely -> explicitly

zeke commented on docs/api/app.md:
```diff
@@ -527,6 +527,151 @@ Adds `tasks` to the [Tasks][tasks] category of the JumpList on Windows.
+Sets or removes a custom Jump List for the application, and returns one of the
+following strings:
+
+* `"ok"` - Nothing went wrong.
```
I would suggest dropping the doublequotes on these strings for cleanliness. It's clear enough that they are strings.

zeke commented on docs/api/app.md:
```diff
@@ -527,6 +527,151 @@ Adds `tasks` to the [Tasks][tasks] category of the JumpList on Windows.
+`JumpListCategory` objects should have the following properties:
+
+* `type` String - One of the following:
+  * `"tasks"` - Items in this category will be placed into the standard `Tasks`
```
Same here. No quotes.

enlight commented:
Updated the docs as suggested by @zeke

zcbenz commented:
Thanks!

bpasero commented:
@enlight thanks for this API! If I am calling into setJumpList is it still possible to add recent files via `app.addRecentDocument`?

bpasero commented:
@enlight actually now I see the docs and there is a category `recent` and `frequent`. I assume the category recent is needed so that `app.addRecentDocument` works. 

Does windows fill the `frequent` entries automatically based on usage of files?

enlight commented:
@bpasero Yes, files added via `app.addRecentDocument` should show up in the `recent` category, I believe files selected via `dialog.showOpenDialog` will also end up in that list. The `frequent` category is populated in the same way. Here's the relevant text from the docs:

> The contents of the Frequent and Recent categories are calculated for each application that uses SHAddToRecentDocs directly. In some cases of user action, such as opening a file through Windows Explorer or using the common file dialog box to open, save, or create a file, the Shell calls SHAddToRecentDocs on behalf of an application and those calls are also taken into account in the usage statistics. The Shell also calls SHAddToRecentDocs on behalf of the application when a destination is launched from its Jump List. However, it is good practice for the application to explicitly call SHAddToRecentDocs itself even if it is expected that the Shell will make the call. This guarantees that the usage is recorded, and the algorithms for tracking recent or frequent usage will correct for any duplicate calls.

bpasero commented:
@enlight great. one thing puzzles me though: we (VS Code) never could get the entries we added via `app.addRecentDocument` to show up in the task bar. Now, when I use the new `setJumpList` and add an empty category `recent`, they start to show up! This was driving me crazy and I spent hours over hours to find the bug on our end by checking if some Windows registry key was wrong or missing...

One thing I noticed is that the issue was caused by us calling `app.setUserTasks`. With the new `setJumpList` I am not using that call anymore and if I take it out completely (and also do not call `app.setJumpList`, the recent file entries start to show up.

Anyhow, now with `setJumpList` it seems to work just fine 👍

enlight commented:
@bpasero The default Jump List displays the `Recent` category, but a custom Jump List completely replaces the default one (rather than amending it), so all the categories you want to show up must be added explicitly. `app.setUserTasks` creates a custom Jump List that only contains the `Tasks` category, which is why there is no `Recent` category displayed in the Jump List after you call that function.

Anyway, happy to hear you now have one less thing to worry about :)

bpasero commented:
@enlight that explains everything :). I think it should be documented though 👍

bpasero commented:
@enlight have you heard about an issue where the custom entries of the jump list are all gone once you do this:

![image](https://cloud.githubusercontent.com/assets/900690/20066840/53c3ddd8-a513-11e6-8e7a-82a37649117e.png)

As soon as I pick "Remove from this list", the custom category is gone for good. No restart helps. The only thing that does help is a complete reinstall :-(

enlight commented:
@bpasero Did a search through the VSCode repo and didn't see [`app.getJumpListSettings()`](https://github.com/electron/electron/blob/master/docs/api/app.md#appgetjumplistsettings-windows) called anywhere, so...

When a user unpins an item from the Jump List Windows will keep track of it, if you then attempt to add that item again with `app.setJumpList()` Windows will simply throw out the whole custom category the item belongs to.

To fix this in VSCode you need to call `app.getJumpListSetting()` to obtain the `removedItems` list before calling `app.setJumpList()`, and ensure that none of the of the items in `removedItems` are added back into the Jump List. Note that this doesn't mean that once removed an item can never be added back into the Jump List, Windows will clear the `removedItems` list after a successful call to `app.setJumpList()`, at which point you can call `app.setJumpList()` again to add the previously removed item back in.

bpasero commented:
@enlight thanks man, I was not aware of this method. will give it a try!!
